### PR TITLE
[PTX-20636] TC Automation - ResizePVCToMaxLimit

### PR DESCRIPTION
**Pull request by kphalgun-px**
**What this PR does / why we need it**:
This PR adds "ResizePVCToMaxLimit" test case to validate volume resize, ensuring they expand correctly to their maximum limits.

**Which issue(s) this PR fixes** (optional)
Closes #PTX-20636,PTX-20637

**Special notes for your reviewer**:
~~Please review the Jenkins build for the test case "ResizePVCToMaxLimit" for app "fio-fa-cloud-drives" with FACD volumes [PASS]~~

~~https://jenkins.pwx.dev.purestorage.com/job/Users/job/Leela/job/tp-byoc/190/~~

~~Aetos Dashboard: https://aetos.pwx.purestorage.com/resultSet/testSetID/396040~~

~~Screenshot 2023-10-26 at 3 06 16 AM~~

~~Please review the Jenkins build for the test case "ResizePVCToMaxLimit" for app "fio-fa-davol" with FADA volumes [PARTIAL SUCCESS]~~

~~https://jenkins.pwx.dev.purestorage.com/job/Users/job/Leela/job/tp-byoc/191/~~

~~Aetos Dashboard: https://aetos.pwx.purestorage.com/resultSet/testSetID/396053~~

~~Screenshot 2023-10-26 at 3 18 41 AM~~

~~Screenshot 2023-10-26 at 3 27 11 AM~~

~~Please review the Jenkins build for the test case "ResizePVCToMaxLimit" for app "fio" with VsCD volumes [PARTIAL SUCCESS]~~

~~https://jenkins.pwx.dev.purestorage.com/job/Users/job/Leela/job/tp-byoc/194/~~

~~Aetos Dashboard: https://aetos.pwx.purestorage.com/resultSet/testSetID/396084~~

~~Screenshot 2023-10-26 at 3 54 39 AM~~

Please review the Jenkins build for the test case "ResizePVCToMaxLimit" for app "fio-fa-davol" with FADA volumes [PASS]

https://jenkins.pwx.dev.purestorage.com/job/Users/job/Leela/job/tp-byoc/242/

Aetos Dashboard: https://aetos.pwx.purestorage.com/resultSet/testSetID/407630

Please review the Jenkins build for the test case "ResizePVCToMaxLimit" for app "fio-fa-cloud-drives" with FACD volumes [PASS]

https://jenkins.pwx.dev.purestorage.com/job/Users/job/Leela/job/tp-byoc/250/

Aetos Dashboard: https://aetos.pwx.purestorage.com/resultSet/testSetID/407786

### CODE REVIEW

File: tests/basic/pure_test.go [hunk 1]
```diff
@@ -335,3 +338,116 @@ var _ = Describe("{BringUpLargePodsVerifyNoPanic}", func() {
...
+		volumeMap           = make(map[string][]*api.Volume)
+		minVolSizeIncrement = uint64(1 * units.TiB)
+	)
+
+	// getMaxVolSize gets the maximum volume size based on the given volType
+	getMaxVolSize := func(volType string) uint64 {
+		switch volType {
+		case "FACD":
+			return 40 * units.TiB
+		default:
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:358]: check the max limit of other types like normal vsphere and make that as default

File: tests/basic/pure_test.go [hunk 2]
```diff
@@ -335,3 +338,116 @@ var _ = Describe("{BringUpLargePodsVerifyNoPanic}", func() {
...
+	// getNextSize gets the next closest multiple in powers of 2 greater than currentSize
+	getNextSize := func(currentSize uint64) uint64 {
+		if currentSize < minVolSizeIncrement {
+			return minVolSizeIncrement
+		}
+		return currentSize * 2
+	}
+
+	// resizeVolume resizes given the volume using powers of 2 until the max limit is reached
+	resizeVolume := func(volType string, vol *api.Volume) error {
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:372]: put these functions inside It

Comment (kphalgun-px) [reply tests/basic/pure_test.go]: Done.

File: tests/basic/pure_test.go (hunk 1 above)
Comment (kphalgun-px) [reply tests/basic/pure_test.go]: Done.

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

File: tests/basic/pure_test.go [hunk 3]
```diff
@@ -1014,3 +1017,285 @@ var _ = Describe("{RestartPXWhileVolCreate}", func() {
...
+	var (
+		contexts            = make([]*scheduler.Context, 0)
+		backend             = BackendUnknown
+		volumeMap           = make(map[VolumeType][]*api.Volume)
+		volumeCtxMap        = make(map[string]*scheduler.Context)
+		steps        uint64 = 5
+	)
+
+	JustBeforeEach(func() {
+		StartTorpedoTest("ResizePVCToMaxLimit", "Validate PVC resize to max limit", nil, 72657582)
```
Comment (mkoppal-px) [inline tests/basic/pure_test.go:1057]: Is this even a valid testrail ID?

Comment (kphalgun-px) [reply tests/basic/pure_test.go:1057]: https://portworx.testrail.net/index.php?/tests/view/72657582

**Review by mkoppal-px — APPROVED** [review]
LGTM except for one minor comment

File: tests/basic/pure_test.go (hunk 3 above)
Comment (mkoppal-px) [reply tests/basic/pure_test.go:1057]: This is incorrect. The link you have mentioned is from a test run not the actual test case. The correct link is this - https://portworx.testrail.net/index.php?/cases/view/87940

Comment (kphalgun-px) [reply tests/basic/pure_test.go]: Done. Thank you.

**Review by vprabhakar-px — APPROVED** [review]
LGTM

File: drivers/volume/portworx/portworx.go [hunk 4]
```diff
@@ -1258,6 +1258,13 @@ func (d *portworx) ValidateCreateVolume(volumeName string, params map[string]str
 				Cause: fmt.Sprintf("Volume has invalid state. Actual:%v", vol.State),
 			}
 		}
+
+		// TODO: remove this retry once PWX-27773 is fixed
+		// DevicePath
```
[the remark this answers is not in the document]
Comment (kphalgun-px) [reply drivers/volume/portworx/portworx.go:1262]: Done.

**Pull request merged**